storage/src/journal/fixed.rs

```diff
@@ -306,33 +370,31 @@ impl<E: Storage + Metrics, A: Codec<Cfg = ()> + FixedSize> Journal<E, A> {
             std::cmp::Ordering::Less => {}
         }
         let rewind_to_blob_index = journal_size / self.cfg.items_per_blob;
-        if rewind_to_blob_index < *self.oldest_blob().0 {
+        if rewind_to_blob_index < self.oldest_blob_index() {
             return Err(Error::InvalidRewind(journal_size));
         }
         let rewind_to_offset = (journal_size % self.cfg.items_per_blob) * Self::CHUNK_SIZE_U64;
-        let mut current_blob_index = *self.newest_blob().0;
 
         // Remove blobs until we reach the rewind point.
-        while current_blob_index > rewind_to_blob_index {
-            let blob = match self.blobs.remove(&current_blob_index) {
-                Some(blob) => blob,
-                None => return Err(Error::MissingBlob(current_blob_index)),
-            };
-            blob.close().await?;
-            self.context
-                .remove(&self.cfg.partition, Some(&current_blob_index.to_be_bytes()))
-                .await?;
-            debug!(blob = current_blob_index, "unwound over blob");
-            self.tracked.dec();
-            current_blob_index -= 1;
+        if rewind_to_blob_index < self.tail_index {
+            let mut current_blob_index = self.tail_index - 1;
+            while current_blob_index > rewind_to_blob_index {
+                let Some(blob) = self.blobs.remove(&current_blob_index) else {
+                    return Err(Error::MissingBlob(current_blob_index));
+                };
+                self.remove_blob(current_blob_index, blob).await?;
+                current_blob_index -= 1;
+            }
         }
 
-        // Truncate the rewind blob to the correct offset.
-        let rewind_blob = match self.blobs.get_mut(&rewind_to_blob_index) {
-            Some(blob) => blob,
-            None => return Err(Error::MissingBlob(rewind_to_blob_index)),
-        };
-        rewind_blob.resize(rewind_to_offset).await?;
+        // Set up the new tail blob and truncate it to the correct offset.
+        if rewind_to_blob_index != self.tail_index {
```

Comment:
I think we need to remove tail first (like before)? Otherwise, a shutdown could lead to a gap?

Reply:
Yes, wasn't thinking through failure conditions here, and didn't adequately simplify it once I removed the Immutable wrapper. I have fixed the issue, added a comment that it's mandatory to go in reverse order so it's not re-introduced, and simplified the code.